sw/device/lib/dif/dif_spi_host_unittest.cc

```diff
@@ -0,0 +1,517 @@
+                 {
+                     {SPI_HOST_CONFIGOPTS_CLKDIV_0_OFFSET, 1},
+                     {SPI_HOST_CONFIGOPTS_CSNIDLE_0_OFFSET, 0},
+                     {SPI_HOST_CONFIGOPTS_CSNTRAIL_0_OFFSET, 0},
+                     {SPI_HOST_CONFIGOPTS_CSNLEAD_0_OFFSET, 0},
+                     {SPI_HOST_CONFIGOPTS_FULLCYC_0_BIT, false},
+                     {SPI_HOST_CONFIGOPTS_CPHA_0_BIT, false},
+                     {SPI_HOST_CONFIGOPTS_CPOL_0_BIT, false},
+                 });
+  EXPECT_WRITE32(SPI_HOST_CONTROL_REG_OFFSET,
+                 {
+                     {SPI_HOST_CONTROL_SPIEN_BIT, true},
+                 });
+
+  EXPECT_EQ(dif_spi_host_configure(&spi_host_, config_), kDifOk);
+}
+
+// Checks that the chip select options get written to the appropriate fields in
+// the config register.
+TEST_F(ConfigTest, ChipSelectOptions) {
+  config_.chip_select.idle = 1;
+  config_.chip_select.trail = 2;
+  config_.chip_select.lead = 3;
+
+  ExpectDeviceReset();
+  EXPECT_WRITE32(SPI_HOST_CONFIGOPTS_REG_OFFSET,
+                 {
+                     {SPI_HOST_CONFIGOPTS_CLKDIV_0_OFFSET, 0},
+                     {SPI_HOST_CONFIGOPTS_CSNIDLE_0_OFFSET, 1},
+                     {SPI_HOST_CONFIGOPTS_CSNTRAIL_0_OFFSET, 2},
+                     {SPI_HOST_CONFIGOPTS_CSNLEAD_0_OFFSET, 3},
+                     {SPI_HOST_CONFIGOPTS_FULLCYC_0_BIT, false},
+                     {SPI_HOST_CONFIGOPTS_CPHA_0_BIT, false},
+                     {SPI_HOST_CONFIGOPTS_CPOL_0_BIT, false},
+                 });
+  EXPECT_WRITE32(SPI_HOST_CONTROL_REG_OFFSET,
+                 {
+                     {SPI_HOST_CONTROL_SPIEN_BIT, true},
+                 });
+
+  EXPECT_EQ(dif_spi_host_configure(&spi_host_, config_), kDifOk);
+}
+
+// Checks that the SPI cycle, polarity and phase options get written to the
+// appropriate fields in the config register.
+TEST_F(ConfigTest, SpiOptions) {
+  config_.full_cycle = true;
+  config_.cpol = true;
+  config_.cpha = true;
+
+  ExpectDeviceReset();
+  EXPECT_WRITE32(SPI_HOST_CONFIGOPTS_REG_OFFSET,
+                 {
+                     {SPI_HOST_CONFIGOPTS_CLKDIV_0_OFFSET, 0},
+                     {SPI_HOST_CONFIGOPTS_CSNIDLE_0_OFFSET, 0},
+                     {SPI_HOST_CONFIGOPTS_CSNTRAIL_0_OFFSET, 0},
+                     {SPI_HOST_CONFIGOPTS_CSNLEAD_0_OFFSET, 0},
+                     {SPI_HOST_CONFIGOPTS_FULLCYC_0_BIT, true},
+                     {SPI_HOST_CONFIGOPTS_CPHA_0_BIT, true},
+                     {SPI_HOST_CONFIGOPTS_CPOL_0_BIT, true},
+                 });
+  EXPECT_WRITE32(SPI_HOST_CONTROL_REG_OFFSET,
+                 {
+                     {SPI_HOST_CONTROL_SPIEN_BIT, true},
+                 });
+
+  EXPECT_EQ(dif_spi_host_configure(&spi_host_, config_), kDifOk);
+}
+
+class TransactionTest : public SpiHostTest {
+ protected:
+  MockFifo fifo_;
+};
+
+// Checks that an opcode segment is sent correctly.
+TEST_F(TransactionTest, IssueOpcode) {
+  dif_spi_host_segment segments[] = {
```

Comment:
You can use `std::array<>` to avoid needing to use ARRAYSIZE.

Reply:
I _could_, but C++ doesn't allow designated initializers for std::array.  I judge the readability of the initializer above the unpleasantness of using a macro.

`std::array<>` also appears to require you specify the size of the array as a template argument, whereas the C-style declaration permits the array to be of unspecified size to be computed by the compiler.

Reply:
_shrug_

Note that C++ doesn't have array designated inits so it's moot either way. =P